checker/src/org/checkerframework/checker/optional/OptionalVisitor.java

```diff
@@ -0,0 +1,340 @@
+ * Framework Manual.
+ *
+ * @checker_framework.manual #optional-checker Optional Checker
+ */
+public class OptionalVisitor
+        extends BaseTypeVisitor</* OptionalAnnotatedTypeFactory*/ BaseAnnotatedTypeFactory> {
+
+    private final TypeMirror collectionType;
+
+    public OptionalVisitor(BaseTypeChecker checker) {
+        super(checker);
+        collectionType = types.erasure(TypesUtils.typeFromClass(Collection.class, types, elements));
+    }
+
+    /** Provides a way to query the Nullness Checker. */
+    NullnessAnnotatedTypeFactory getNullnessTypeFactory() {
+        return checker.getTypeFactoryOfSubchecker(NullnessChecker.class);
+    }
+
+    @Override
+    protected BaseTypeValidator createTypeValidator() {
+        return new OptionalTypeValidator(checker, this, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.get */
+    private boolean isCallToGet(ExpressionTree expression) {
+        return OptionalUtils.isMethodInvocation(expression, "get", 0, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.isPresent */
+    private boolean isCallToIsPresent(ExpressionTree expression) {
+        return OptionalUtils.isMethodInvocation(expression, "isPresent", 0, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to java.util.Optional.of */
+    private boolean isCallToOf(ExpressionTree expression) {
+        return OptionalUtils.isMethodInvocation(expression, "of", 1, atypeFactory);
+    }
+
+    /** @return true iff expression is a call to Optional creation: of, ofNullable. */
+    private boolean isOptionalCreation(MethodInvocationTree methInvok) {
+        return OptionalUtils.isMethodInvocation(methInvok, "of", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "ofNullable", 1, atypeFactory);
+    }
+
+    /**
+     * @return true iff expression is a call to Optional elimination: get, orElse, orElseGet,
+     *     orElseThrow.
+     */
+    private boolean isOptionalElimation(MethodInvocationTree methInvok) {
+        return OptionalUtils.isMethodInvocation(methInvok, "get", 0, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElse", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElseGet", 1, atypeFactory)
+                || OptionalUtils.isMethodInvocation(methInvok, "orElseThrow", 1, atypeFactory);
+    }
+
+    @Override
+    public Void visitConditionalExpression(ConditionalExpressionTree node, Void p) {
+        handleTernaryIsPresentGet(node);
+        return super.visitConditionalExpression(node, p);
+    }
+
+    /**
+     * Part of rule #3.
+     *
+     * <p>Pattern match for: {@code VAR.isPresent() ? VAR.get().METHOD() : VALUE}
+     *
+     * <p>Prefer: {@code VAR.map(METHOD).orElse(VALUE);}
+     */
+    public void handleTernaryIsPresentGet(ConditionalExpressionTree node) {
```

Comment:
This logic for `isPresent` followed by `get` would more generally be handled by a transfer function, instead of coding the pattern on Trees for ternaries and ifs separately.
As the current approach works fine for the tests, I'm fine with merging it. But this would be a good future generalization.